### ISSUE TITLE
kern.maxfile not being used / too many open files

### ISSUE THREAD

**Issue by NorseGaud** [docker/for-mac#4950]
- [X] I have tried with the latest version of my channel (Stable or Edge)
  - [N/A] I have uploaded Diagnostics
  - Diagnostics ID: N/A

### Information
Per https://github.com/docker/for-mac/issues/1009#issuecomment-264874864, I should be able to edit the max connections, however, I don't see the directory indicated.
  - macOS Version: 10.15.7

Running GitLab on macOS/Docker Desktop within a few jobs being run starts throwing `too many open files` errors. I've opened https://gitlab.com/gitlab-org/gitlab/-/issues/255992 to see if they have any suggestions, but I'd also like to rule out docker as the cause since running the gitlab container on linux doesn't seem to have the issue.

**Comment by NorseGaud** [comment]
@djs55 , do you know how to achieve this with the latest edge version?

**Comment by djs55** [comment]
@NorseGaud there was a change in Docker Desktop edge 2.3.7.0 (which has now propagated to stable 2.4.0.0), from https://docs.docker.com/docker-for-mac/edge-release-notes/#docker-desktop-community-2370

> Removed an artificial file descriptor limit (setrlimit) of 10240. Docker Desktop now relies on the kernel to impose limits via kern.maxfiles and kern.maxfilesperproc.

For a long time we used `setrlimit` to set a limit of 10240 file descriptors but this wasn't enough for some use-cases (!) so we bumped up the soft limit to a very high value:

```
$ cd ~/Library/Containers/com.docker.docker/Data/log/host/
$ grep "Maximum number of file" *
com.docker.supervisor.log:time="2020-10-02T10:50:07+01:00" level=info msg="Maximum number of file descriptors is 9223372036854775807"
```

which means the real limit should now be the macOS kernel one:
```
$ sysctl -a | grep kern.maxfile
kern.maxfiles: 49152
kern.maxfilesperproc: 24576
```
It's possible this isn't working somehow. When you're seeing too many open files errors, could you check with `lsof -p <pid>` on the Mac (or Activity Monitor -> select process -> click on (info) -> Open Files and Ports) to see if you can spot where the bottleneck might be? I'm suspicious of the `com.docker.backend` process since it contains the fileserver. It's also worth checking `com.docker.hyperkit` (VM virtual hardware) and `com.docker.vpnkit` (networking)

**Comment by NorseGaud** [comment]
Thanks for the reply and details! I'll run the test of lsof in a little bit. Here is what I see on my laptop in the meantime:

```
❯ grep "Maximum number of file" *
com.docker.supervisor.log:time="2020-09-09T10:27:10-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-10T12:53:36-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-10T13:57:25-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-10T14:00:23-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-10T14:06:20-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-10T14:15:40-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-15T12:12:24-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-16T11:17:07-04:00" level=info msg="Maximum number of file descriptors is 10240"
com.docker.supervisor.log:time="2020-09-18T14:41:22-04:00" level=info msg="Maximum number of file descriptors is 10240"

❯ sysctl -a | grep kern.maxfile
kern.maxfiles: 98304
kern.maxfilesperproc: 49152
```

**Comment by NorseGaud** [comment]
When three jobs are running:

```
Every 1.0s: for pid in $(pgrep docker); do ...  Veertu.local: Fri Oct  2 15:08:54 2020

COMMAND /Library/PrivilegedHelperTools/com.docker.vmnetd
      15
COMMAND /Applications/Docker.app/Contents/MacOS/com.docker.supervisor
      41
COMMAND /Applications/Docker.app/Contents/MacOS/com.docker.backend
      65
COMMAND com.docker.osxfs
    6683
COMMAND com.docker.vpnkit
      26
COMMAND com.docker.driver.amd64-linux
      32
COMMAND docker
      13
COMMAND com.docker.hyperkit
      45
```

osxfs seems to climb ~400 open files each time a job finishes or completes.

I then switched my edge docker desktop back to gRPC FUSE and tried again. This time just queuing up jobs and not letting them actually run (be picked up by a gitlab runner). Same result, gitlab starts throwing too many open files and here is the lsof:

```
Every 1.0s: for pid in $(pgrep docker); do ...  Veertu.local: Fri Oct  2 15:25:18 2020

COMMAND /Library/PrivilegedHelperTools/com.docker.vmnetd
      15
COMMAND /Applications/Docker.app/Contents/MacOS/com.docker.supervisor
      41
COMMAND /Applications/Docker.app/Contents/MacOS/com.docker.backend
    7170
COMMAND com.docker.osxfs
      14
COMMAND com.docker.vpnkit
      26
COMMAND com.docker.driver.amd64-linux
      31
COMMAND docker
      13
COMMAND com.docker.hyperkit
      31
```

**Comment by NorseGaud** [comment]
@djs55 let me know if you need anything else for this. More than happy to jump on a google meet/screen share too.

**Comment by NorseGaud** [comment]
Wanted to bump this issue. We have a few people using docker desktop for mac receiving this problem when trying to use gitlab's container tag to run a local gitlab. It's becoming painful for them to restart the container anytime it crashes.

**Comment by NorseGaud** [comment]
Hey @djs55, is there anything I can do to help expedite this? More than happy to help troubleshoot/test.

**Comment by NorseGaud** [comment]
Bump 2021. Happy new year

**Comment by NorseGaud** [comment]
Hi @stephen-turner and @djs55, is there anything that can be done to increase the priority of this? It's a major blocker for large apps running on macOS docker. Gitlab for example is un-usable after a few minutes due to this issue.

**Comment by stephen-turner** [comment]
Sorry not to give you the answer you want, but you seem to be the only person who's reported running into this, so it's not high up the queue at the moment.

**Comment by NorseGaud** [comment]
Still happening on 3.3.1

**Comment by NorseGaud** [comment]
@stephen-turner hello good sir! Could you reopen this? I can't seem to find a way to reopen it

**Comment by NorseGaud** [comment]
I think I finally found the issue! It seems like the limits must be set by a LaunchDaemon on boot. Here is what I saw:

```
❯ sudo ulimit -n
256
```

Trying to change that and even reboot didn't seem to help. It always seemed to be 256.

So, what I did was setup a plist file on boot to set the values and then rebooted. So far it hasn't crashed! I've run a ton of CI jobs in gitlab/docker and it's all working great. 

Here is the plist: https://gist.github.com/NorseGaud/4e7cbc53f3a31a07179ca9cf5870924e
